### ISSUE TITLE
what is bad.list used for?

### ISSUE THREAD

**Issue by reachsumit** [pjreddie/darknet#979]
I am trying to train yolo on my custom dataset. However, the final predictions.jpg file is always empty for me.

- There is file in my root folder called bad.list, that contains thousands of entries such as:

> data/labels/32_0.png
> data/labels/33_0.png
> data/labels/34_0.png
> data/labels/35_0.png
> ...

- At the time of prediction, I get similar errors:

> Cannot load image "data/labels/32_0.png"
> Cannot load image "data/labels/33_0.png"
> Cannot load image "data/labels/34_0.png"
> Cannot load image "data/labels/35_0.png"
> Cannot load image "data/labels/36_0.png"
> ...

These png images are not part of my training dataset and do not exist in any configuration file.
I'm guessing this is might be the reason why my predictions are empty with no bounding box around the desired object(?).

Could anybody please throw some light on this.

**Comment by timombre** [comment]
data/labels/ is a folder containing the "labels" used to write the text of the bounding box.
It is called by the function load_alphabet (in turn called by draw_detections ) in src/images.c
You should copy the folder in the directory where you train your model.

**Comment by timombre** [comment]
The folder label's name is somewhat confusing, it does not correspond to labels from the network, it is just a display tool for the boxes, it has no connection whatsoever to your dataset. If you go to darknet/data/label, you will see them, they are just a bunch of characters turned in images by the script make labels.py

Maybe you included those images in you training or testing files (hence it end up in bad.list), this would be a mistake (check your script generating the path of the files).
If you only have one class, you may have issues solved with this fix: https://github.com/pjreddie/darknet/pull/320/commits/7ee9ebda1f872ef18b67b1a984af2f84fb0d70c3

**Comment by timombre** [comment]
Have you correctly set up your path in voc.data .  does it print the line 

printf("%s: Predicted in %f seconds.\n", input, what_time_is_it_now()-time);

before generating the picture ?

**Comment by reachsumit** [comment]
Yes, I see below:

> data/test.jpg: Predicted in 0.022667 seconds.

If I use -thresh to be very low (0.001), then the output is:
data/test.jpg: Predicted in 0.022667 seconds.
building: 8%
building: 0%
building: 0%
building: 0%

And in the prediction, the single box is way off any target, as you can see here: https://i.imgur.com/U3OHiZi.png

Also, I reverse engineered the bounding boxes from my yolo training boxes. The training set images look like this: https://imgur.com/a/vCyZeYu

As you can see there are many overlapping boxes. Would Yolo be a good approach here? Could overlapping boxes be the reason for failure?

**Comment by timombre** [comment]
in detector.c  add the commands  in  void test_detector
save_image(im, "before");
save_image(sized, "whatnetsees");
before the the calling of draw_detections
check that it is correct

**Comment by reachsumit** [comment]
Thank you!

Unfortunately, I can't edit the source as I am running this task on my University's GPU cluster. They already installed the latest darknet and I interact with it using the darknet commands along with my data and config files.

Could there be another way?

**Comment by reachsumit** [comment]
Here's another output, at -thresh 0: https://i.imgur.com/oMpNvtv.png
Each image resolution is about 600x600 and 800x800 (RGB).

**Comment by timombre** [comment]
Maybe you could try locally on CPU for a single image (should take around 8 s but it's ok for testing).

**Comment by reachsumit** [comment]
The sample that I showed you was recreated from the bounding box data that I gave yolo. I drew those boxes to show you how the input might look like.

My setup is this:
In my /data/ directory I have full-sized(650x650) training images placed **without any bounding boxes** on them. In the same directory I have .txt files, one for each training jpeg image, having the same name as the corresponding training image. The .txt file contains the bounding box information for the objects inside the corresponding image.

So, to answer your question, no, I am not manually extracting every red block in a picture of its own.

**Comment by reachsumit** [comment]
You Good Sir, have saved me hours of debugging and frustration.

I downloaded the trained weights on my local CPU and did predictions the same way as you guided me to. The predictions is very accurate. (prediction result: https://i.imgur.com/Of9V2qg.png)
before image: https://i.imgur.com/6DFb76x.png
whatnetsees: https://i.imgur.com/OuuEuxq.png

I don't know why my University's GPU node is unable to predict this way.

**Comment by timombre** [comment]
Glad I have been useful. You should try with a local GPU install to see if it is a GPU issue. What you can do also is to try to use the -nogpu option on your GPU cluster and see if it still bugs, if it is the case, ask the cluster manager to fix the darknet install (does it work with the yolov3 standard on your GPU cluster ?).

**Comment by reachsumit** [comment]
Will do! Thank you. :+1:

**Comment by skipperkongen** [comment]
@reachsumit you thank @timombre but what advice did you follow to get predictions? I had the same issues with Cannot load image "data/labels/32_0.png" and fixed them by copying the folder into the the location where I train my customer images. That fixed that problem. I also added the debug lines to detector.c, which is great. But after that, what did you do? Run on CPU only?

**Comment by caohe115** [comment]
Can you summarize the solution? I don't understand

**Comment by AMITKUMARMALLICK18** [comment]
@reachsumit , hi
How did you solve this bad.list? I am not getting this. Can you please help me?

**Comment by Yuji-github** [comment]
I think I solved the issues. 
Assume you are working on Google Colab. 
Also, you have made a custom training model in your Google Drive somewhere. 

Ex) /content/drive/MyDrive/YOLO_Custom   
YOLO_Custom has your model (backup, train, test, folders that you uploaded or created) 

In the same location, you have your own 'data' folder that contains obj.data and obj.names files. 
However, you do not have a 'labels' folder in the 'data' folder. 

Here are the simple solutions:
Go to the 'original' darknet -> data file 
Download labels file on your PC. 
Then, upload the labels files into your own 'data' folder. 

The reason you got the error is, YOLO was trying to find labels file from the current location where you created your own data folder. 
But, there is no labels folder inside of data folder. So, you got the erros

**Comment by bibekjit** [comment]
@Yuji-github I had the same issue in google colab. But I fixed it by uploading my custom_data into darknet/data along with obj.data, obj.names, train.txt and test.txt. But when I ran it locally, I moved the custom_data folder and all the files mentioned in darknet/data but I still got a bounding box with no class name. I ran the code on jupyter notebook on ubuntu. Could you help me with this

**Comment by Yuji-github** [comment]
@bibekjit I understand that you have the bad.list on your local computer (Linux.) 
If I were you, I will make sure obj.data correctly specifies the file path. 

Ex) obj.data
train = your_local_path/train.txt

Also, make sure the bounding box (annotation) is encoded by JSON (dumps) before saving. 
Then, you write it as a txt file.

I hope it helps, and wish you luck.

**Comment by bibekjit** [comment]
@Yuji-github thanks for the advice, but I fixed this error. All I had to do was make the darknet dir as my cwd before running the detector test.
